### ISSUE TITLE
Feature request: proxy chains

### ISSUE THREAD

**Issue by poteyto** [darkk/redsocks#18]
Hello,

It would be great if redsocks could support proxy chaining... something  like this would be really cool:

redsocks {
...
via {
ip
port
type
via{...}
}
}

**Comment by darkk** [comment]
I think that workaround suggested by @L4R5 is good enough _to access your own private infrastructure_. Yes, it has some performance penalty, but it's not that high (and it's probably even lower with `splice()` support in v0.5) for this specific use-case.

I'm closing the issue as I have no motivation to implement proxy chaining in foreseeable future within redsocks itself.

**Comment by fatality14** [comment]
It isn't that clear how to set up proxy chain with redsocks, especially if you aren't that good in iptables configuration.
So I considered to provide a sample config for people who will face the same problem, cause there isn't that much ways to create proxy chain in linux except proxychains program.

As already been said it is necessary to create two chains similar to one specified in redsocks readme. The main point is that packages sent through REDSOCKS chain are being sent to OUTPUT chain by redsocks even if it can not connect to destination ip of our proxy. The trick is to redirect these packages to e.g. REDSOCKS1 chain (which would send it to another proxy) using destination filtering. The proxy chain might have unlimited number of such package transitions.

So the config for http/https ports is being set this way with iptables-restore command:
> *nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [912:64780]
:POSTROUTING ACCEPT [2079:134800]
:REDSOCKS - [0:0]
:REDSOCKS1 - [0:0]
-A OUTPUT -p tcp -m tcp --dport 443 -j REDSOCKS
-A OUTPUT -p tcp -m tcp --dport 80 -j REDSOCKS
-A OUTPUT -d /second proxy external ip/ -p tcp -m tcp --dport /port/ -j REDSOCKS1
-A REDSOCKS -p tcp -j DNAT --to-destination 127.0.0.1:/redsocks second proxy port/
-A REDSOCKS1 -p tcp -j DNAT --to-destination 127.0.0.1:/redsocks first proxy port/
COMMIT

And the redsocks.conf might look like this:
> base {
	log_debug = on;
	log_info = on;
	log = "stderr";
	daemon = off;
	redirector = iptables;
        rlimit_nofile = 500000;
        redsocks_conn_max = 655650;
}
redsocks {
	local_ip = 127.0.0.1;
	local_port = /redsocks first proxy port/;
	ip = /first proxy external ip/;
	port = /port/;
	type = /your type here/; //socks4, socks5, http-connect, http-relay
	login = "admin";
	password = "1234";
}
redsocks {
	local_ip = 127.0.0.1;
	local_port = /redsocks second proxy port/;
	ip = /second proxy external ip/;
	port = /port/;
	type = /your type here/; //socks4, socks5, http-connect, http-relay
	login = "admin";
	password = "1234";
}

In other words packages are being transferred in the next order on local machine:

1. From OUTPUT chain to REDSOCKS chain
2. From REDSOCKS to redsocks second local proxy (which might have no access to its external proxy)
3. From redsocks second local proxy to OUTPUT (with destination address of second external proxy ip)
4. From OUTPUT to REDSOCKS1 (filtered by destination of second external proxy)
5. From REDSOCKS1 to redsocks first local proxy
6. From redsocks first local proxy to first external proxy
